File: apps/bridge/src/environment.ts
/**
 * Probe local grok CLI and login state for UI banners and pre-handshake checks.
 * Does not read API key contents; only whether usable credentials exist.
 */

import { existsSync } from "node:fs";
import { spawn } from "node:child_process";
import path from "node:path";
import { resolveGrokBin } from "./spawnGrok.js";
import type { EnvironmentInfo } from "./protocol.js";
import { isGrokVersionSupported } from "./versionCheck.js";

/** Default pool capacity (one process per session); override with BRIDGE_POOL_CAPACITY. */
export function poolCapacityFromEnv(): number {
  const raw = Number(process.env.BRIDGE_POOL_CAPACITY ?? 8);
  if (!Number.isFinite(raw) || raw < 1) {
    return 8;
  }
  return Math.min(Math.floor(raw), 16);
}

/**
 * Resolve the ~/.grok/auth.json path (or the Windows user-profile equivalent).
 * @returns Absolute path; if HOME is missing, still returns a path based on an empty string for display.
 */
export function defaultAuthJsonPath(): string {
  const home = process.env.HOME ?? process.env.USERPROFILE ?? "";
  return path.join(home, ".grok", "auth.json");
}

/**
 * Probe auth source: process env XAI_API_KEY first, then local cached_token file.
 * @returns authSource and whether logged in.
 */
export function probeAuthSource(): {
  authed: boolean;
  authSource: EnvironmentInfo["authSource"];
  authPathChecked: string;
} {
  const authPathChecked = defaultAuthJsonPath();
  if (process.env.XAI_API_KEY && process.env.XAI_API_KEY.trim()) {
    return { authed: true, authSource: "xai_api_key", authPathChecked };
  }
  if (existsSync(authPathChecked)) {
    return { authed: true, authSource: "cached_token", authPathChecked };
  }
  return { authed: false, authSource: "none", authPathChecked };
}

/**
 * Run `grok --version`; return null on timeout or failure (soft fail; does not block structural probing).
 * @param bin grok executable path or PATH name.
 * @param timeoutMs Max wait in milliseconds.
 */
export function readGrokVersion(
  bin: string,
  timeoutMs = 3000,
): Promise<string | null> {
  return new Promise((resolve) => {
    const child = spawn(bin, ["--version"], {
      stdio: ["ignore", "pipe", "pipe"],
      env: { ...process.env },
    });
    let out = "";
    const timer = setTimeout(() => {
      child.kill("SIGTERM");
      resolve(out.trim() || null);
    }, timeoutMs);
    child.stdout.on("data", (d: Buffer) => {
      out += d.toString("utf8");
    });
    child.stderr.on("data", (d: Buffer) => {
      out += d.toString("utf8");
    });
    child.on("error", () => {
      clearTimeout(timer);
      resolve(null);
    });
    child.on("close", () => {
      clearTimeout(timer);
      const line = out
        .split("\n")
        .map((s) => s.trim())
        .find(Boolean);
      resolve(line ?? null);
    });
  });
}

/**
 * Aggregate environment probe results.
 * @param poolCapacity Current bridge pool capacity, returned to the UI as well.
 * @returns EnvironmentInfo; ok means CLI found and logged in.
 */
export async function checkEnvironment(
  poolCapacity: number = poolCapacityFromEnv(),
): Promise<EnvironmentInfo> {
  const auth = probeAuthSource();
  let grokPath: string | null = null;
  let version: string | null = null;
  let message = "";

  try {
    grokPath = resolveGrokBin();
  } catch (e) {
    message =
      e instanceof Error
        ? e.message
        : "grok binary not found (set GROK_BIN or install CLI)";
    return {
      grokPath: null,
      version: null,
      authed: auth.authed,
      authSource: auth.authSource,
      authPathChecked: auth.authPathChecked,
      ok: false,
      message,
      poolCapacity,
    };
  }

  version = await readGrokVersion(grokPath);

  const versionCheck = isGrokVersionSupported(version);
  if (!versionCheck.ok) {
    return {
      grokPath,
      version,
      authed: auth.authed,
      authSource: auth.authSource,
      authPathChecked: auth.authPathChecked,
      ok: false,
      message: versionCheck.message,
      poolCapacity,
    };
  }

  if (!auth.authed) {
    message =
      "No grok login detected: run `grok login` or set the XAI_API_KEY environment variable";
    return {
      grokPath,
      version,
      authed: false,
      authSource: "none",
      authPathChecked: auth.authPathChecked,
      ok: false,
      message,
      poolCapacity,
    };
  }

  message = version
    ? `grok ready · ${version} · auth=${auth.authSource}`
    : `grok ready · auth=${auth.authSource}`;

  return {
    grokPath,
    version,
    authed: true,
    authSource: auth.authSource,
    authPathChecked: auth.authPathChecked,
    ok: true,
    message,
    poolCapacity,
  };
}

File: apps/bridge-go/internal/session/environment.go
// Package session owns runtime lifecycle, disk session list, workspace entries,
// and local environment probes for the Go bridge.
package session

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/xai-org/grok-desktop/apps/bridge-go/internal/spawn"
)

// EnvironmentInfo is the CLI / login probe result (secrets never included).
type EnvironmentInfo struct {
	GrokPath        *string `json:"grokPath"`
	Version         *string `json:"version"`
	Authed          bool    `json:"authed"`
	AuthSource      string  `json:"authSource"` // xai_api_key | cached_token | none
	AuthPathChecked string  `json:"authPathChecked"`
	OK              bool    `json:"ok"`
	Message         string  `json:"message"`
	PoolCapacity    int     `json:"poolCapacity"`
}

// PoolCapacityFromEnv reads BRIDGE_POOL_CAPACITY (default 8, max 16).
// @returns Effective capacity in [1, 16]; invalid/empty env falls back to 8.
func PoolCapacityFromEnv() int {
	raw := os.Getenv("BRIDGE_POOL_CAPACITY")
	if raw == "" {
		return 8
	}
	n := 0
	for _, c := range raw {
		if c < '0' || c > '9' {
			return 8
		}
		n = n*10 + int(c-'0')
	}
	if n < 1 {
		return 8
	}
	if n > 16 {
		return 16
	}
	return n
}

// DefaultAuthJSONPath resolves ~/.grok/auth.json.
func DefaultAuthJSONPath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".grok", "auth.json")
}

// ProbeAuthSource checks XAI_API_KEY then local cached_token file.
func ProbeAuthSource() (authed bool, authSource, authPathChecked string) {
	authPathChecked = DefaultAuthJSONPath()
	if key := strings.TrimSpace(os.Getenv("XAI_API_KEY")); key != "" {
		return true, "xai_api_key", authPathChecked
	}
	if st, err := os.Stat(authPathChecked); err == nil && !st.IsDir() {
		return true, "cached_token", authPathChecked
	}
	return false, "none", authPathChecked
}

// ReadGrokVersion runs `grok --version` with a soft timeout.
func ReadGrokVersion(bin string, timeoutMs int) *string {
	if timeoutMs <= 0 {
		timeoutMs = 3000
	}
	cmd := exec.Command(bin, "--version")
	cmd.Env = os.Environ()
	spawn.HideConsoleWindow(cmd)
	done := make(chan string, 1)
	go func() {
		out, err := cmd.CombinedOutput()
		if err != nil && len(out) == 0 {
			done <- ""
			return
		}
		lines := strings.Split(string(out), "\n")
		for _, line := range lines {
			if t := strings.TrimSpace(line); t != "" {
				done <- t
				return
			}
		}
		done <- ""
	}()
	select {
	case v := <-done:
		if v == "" {
			return nil
		}
		return &v
	case <-time.After(time.Duration(timeoutMs) * time.Millisecond):
		_ = cmd.Process.Kill()
		return nil
	}
}

// CheckEnvironment aggregates CLI + login probe results.
// Version support check is T3 on Go bridge — we skip hard version gates and
// only require binary found + auth present for ok=true.
func CheckEnvironment(poolCapacity int) EnvironmentInfo {
	if poolCapacity < 1 {
		poolCapacity = PoolCapacityFromEnv()
	}
	authed, authSource, authPath := ProbeAuthSource()
	bin, err := spawn.ResolveGrokBin()
	if err != nil {
		return EnvironmentInfo{
			GrokPath: nil, Version: nil, Authed: authed, AuthSource: authSource,
			AuthPathChecked: authPath, OK: false,
			Message: err.Error(), PoolCapacity: poolCapacity,
		}
	}
	grokPath := bin
	version := ReadGrokVersion(bin, 3000)
	if !authed {
		return EnvironmentInfo{
			GrokPath: &grokPath, Version: version, Authed: false, AuthSource: "none",
			AuthPathChecked: authPath, OK: false,
			Message:      "No grok login detected: run `grok login` or set the XAI_API_KEY environment variable",
			PoolCapacity: poolCapacity,
		}
	}
	msg := "grok ready · auth=" + authSource
	if version != nil {
		msg = "grok ready · " + *version + " · auth=" + authSource
	}
	return EnvironmentInfo{
		GrokPath: &grokPath, Version: version, Authed: true, AuthSource: authSource,
		AuthPathChecked: authPath, OK: true, Message: msg, PoolCapacity: poolCapacity,
	}
}
